### PR TITLE
[llama4] add auxiliary-loss-free load balancing to MoE token routing

### PR DESCRIPTION
**Pull request by tianyu-l**
There are two issues in this solution:
1. Communication (sync tokens per expert across all DP ranks) happens on the default stream. Maybe need to arrange it on FSDP/DDP comm stream.
2. The communication is blocking experts bias update, thus always exposed.

We need to evaluate if 2 is a problem to performance. 1 is OK if 2 is acceptable.

### CODE REVIEW

File: torchtitan/experiments/llama4/model/moe.py [hunk 1]
```diff
@@ -209,6 +213,35 @@ def __init__(self, model_args: TransformerModelArgs):
...
+            persistent=True,
+        )
+        self.register_buffer(
+            "tokens_per_expert",
+            torch.zeros(num_experts, dtype=torch.float32),
+            persistent=True,
+        )
+
+        # NOTE: forward hook, forward pre hook, or backward pre hook
+        #       would conflict with activation checkpointing
```
Comment (fegin) [inline torchtitan/experiments/llama4/model/moe.py:231]: uh, this make sense that `backward pre hook` would conflict. Didn't think about it through.

**Comment by awgu** [comment]
nit: this expert bias is specifically not a load-balancing _loss_
that is why DeepSeek calls it aux-loss-free load balancing

File: torchtitan/models/llama3/model.py [hunk 2]
```diff
@@ -419,7 +419,7 @@ def __init__(self, model_args: TransformerModelArgs):
 
     def init_weights(
         self,
-        buffer_device: Optional[torch.device] = None,
+        buffer_device: torch.device | None = None,
```
Comment (fegin) [inline torchtitan/models/llama3/model.py:421]: Do you want to remove the import of Optional as well from this file?

File: torchtitan/experiments/llama4/infra/parallelize_llama.py [hunk 3]
```diff
@@ -105,13 +109,36 @@ def parallelize_llama(
...
+    if dp_mesh is not None:
+        # NOTE: Currently this sync is blocking (thus exposed) and happens on the
+        # default compute stream. Need to assess if this is OK performance-wise.
+        def _sync_tokens_per_expert(module, *_):
+            assert isinstance(module, MoE)
+            torch.distributed.all_reduce(
+                module.tokens_per_expert, group=dp_mesh.get_group()
+            )
+
+        for _, transformer_block in model.layers.items():
```
Comment (fegin) [inline torchtitan/experiments/llama4/infra/parallelize_llama.py]: If we are not using the keys, can we change it to `model.layers.values()`?

**Review by fegin — APPROVED** [review]
LGTM, some nits
